### PR TITLE
Update utils.py

### PR DESCRIPTION
**Pull request by LiyuanLucasLiu**
@zdhNarsil updating kl compute via straight-through for:
1. unbiased kl estimation for gradient computation, 
2. support various value estimation method

### CODE REVIEW

File: openrlhf/models/utils.py [hunk 1]
```diff
@@ -8,6 +8,17 @@ def compute_approx_kl(
     log_probs: torch.Tensor,
     log_probs_base: torch.Tensor,
     kl_estimator: str = "k1",
+) -> torch.Tensor:
+    forward_score = compute_approx_kl_forward(logprob, log_probs_base, kl_estimator)
+    backward_score = 0.5 * (logprob - log_probs_base).square()
+
+    return backward_score - backward_score.detach() + forward_score.detach()
```
[the remark this answers is not in the document]
Comment (LiyuanLucasLiu) [reply openrlhf/models/utils.py]: the original implementation is preferred since `backward_score - backward_score.detach()` would result in 0, leading to numerical precision.

**Comment by zdhNarsil** [comment]
# Why $k_3$ is a biased KL estimator and how to fix it


In John Schulman's [blog](http://joschu.net/blog/kl-approx.html), he introduces a KL estimator called '' $k_3$ estimator'' for $\text{KL}(p_\theta \| p_{\text{base}})$:

$$k_3(x) = r(x) - 1 - \log r(x), \quad r(x) = \frac{p_{\text{base}}(x)}{p_\theta(x)}$$

This is a valid estimator because one can easily show:

$$𝔼_{p_\theta(x)}[k_3(x)] = \text{KL}(p_\theta \| p_{\text{base}})$$

For example, the DeepSeek team uses this estimator in their reinforcement learning training to obtain the R1 model.

However, we point out that this provides invalid (i.e., biased) gradient signal for model training; in fact, one can show:

$$
𝔼_{p_\theta(x)}\left[\nabla_\theta k_3(x)\right] = 𝔼_{p_\theta(x)}[\nabla_\theta (\frac{p_{\text{base}}(x)}{p_\theta(x)}) - \nabla_\theta(\log \frac{p_{\text{base}}(x)}{p_\theta(x)}) ] 
$$

$$
= 𝔼_{p_\theta(x)}\left[\nabla_\theta \left(\frac{p_{\text{base}}(x)}{p_\theta(x)}\right) \right] = 𝔼_{p_\theta(x)}\left[-\frac{p_{\text{base}}(x)}{p_\theta(x)} \nabla_\theta \log p_\theta(x) \right] 
$$

$$
\neq 𝔼_{p_\theta(x)}\left[-\log\frac{p_{\text{base}}(x)}{p_\theta(x)} \nabla_\theta \log p_\theta(x) \right] = \nabla_\theta\text{KL}(p_\theta \| p_{\text{base}})
$$

To get an unbiased gradient estimate of $\text{KL}(p_\theta \| p_{\text{base}})$, consider the $k_2$ estimator from the same blog:

$$k_2(x) = \frac{1}{2}\left(\log\frac{p_{\text{base}}(x)}{p_\theta(x)} \right)^2$$

$$\Rightarrow 𝔼_{p_\theta(x)}[\nabla_\theta k_2(x)] = 𝔼_{p_\theta(x)}\left[-\log\frac{p_{\text{base}}(x)}{p_\theta(x)} \nabla_\theta \log p_\theta(x) \right] = \nabla_\theta\text{KL}(p_\theta \| p_{\text{base}})$$

## How to fix

As a result, when we need the value of KL (e.g., to be added to the reward terms), we should use $k_1$ or $k_3$; when we need to compute the gradient of KL (e.g., as a training loss term), we should use $k_2$.

In our [code commit](https://github.com/LiyuanLucasLiu/OpenRLHF/blob/9e5b82199bb817feea81cddb00f6b50cdc404080/openrlhf/models/utils.py#L7-L16), we use a straight-through estimator trick to implement this idea.

**Comment by hijkzzz** [comment]
Why don't we use K2 estimator instead of K3 ?

**Comment by LiyuanLucasLiu** [comment]
> Why don't we use K2 estimator instead of K3 ?

K2 gives biased estimation on the value of KL and unbiased estimation on the gradient of KL. K3 gives biased estimation on the gradient of KL and unbiased estimation of the value of KL. So its better to combine both via ST, which gaves unbiased estimation on both the value and the gradient of KL.
